### ISSUE TITLE
Implement some safeguards

### ISSUE THREAD

**Issue by licaon-kter** [LonamiWebs/Stringlate#42]
Not an urgent issue.

By safeguards I mean having the app inform the user that they might have missed something, eg: end dot, exclamation mark at the end, question mark at the end.

Yes, this can get tricky (if you have them inside the string too), but at least the ones mentioned above look simpler.

**Comment by XyLoNaMiyX** [comment]
And also replacing "..." (three dots) by "…" (ellipsis), or things like " ," by ", ".

Perhaps this could be done *if and only if* the original string has the dot at the end, for example? What about other languages? For example, Spanish uses "¿" and "¡", but it doesn't have to be necessarily at the start of the string. Maybe anywhere in the middle.

**Comment by licaon-kter** [comment]
> if and only if the original string has

Yes, that's a must, again if you put a dot at the end and the original string does not have it, then that needs to be signaled _**"hey, the original string has/does not have X, are you sure you want to save without?"**_

**Comment by XyLoNaMiyX** [comment]
> "hey, the original string has/does not have X, are you sure you want to save without?"

> (eg. Stringlate has >110 OMFG!)

:+1:

**Comment by licaon-kter** [comment]
Umm, that would need some help text anyway, in the Wiki/offline help.

Buuuuttttt, you can add some info:
* below the prev/save/next buttons: *"hey..."*
* or you can add some sort of red-triangle-of-error at the end of the line (or *cough* #39 *cough* multiline ) that you can touch and a toast saying *"hey..."* can appear.

The red triangle of error I have seen in login screens somewhere, when you enter a bad password, not sure it's a simple thing in Android.

**Comment by XyLoNaMiyX** [comment]
> *cough multiline*

I need to look at why the `XmlPullParser` is not reading the multiline. It probably trims the text, and merges ≥ 2 spaces into one.

> The red triangle of error I have seen in login screens somewhere, when you enter a bad password, not sure it's a simple thing in Android.

Yep, it's simple. But I don't know if it's only for errors or also warnings. Although it shouldn't make a difference since I don't really validate anything when updating the translation.

**Comment by licaon-kter** [comment]
That `ellipsis` fits here. :)

**Comment by XyLoNaMiyX** [comment]
> That `ellipsis` fits here.

But most keyboards don't bring this, and if they do (*AnySoftKeyboard* does), people won't necessarily be using it. However, perhaps aggressively replacing automatically is not an option. An "Apply suggested fix" button maybe?

**Comment by licaon-kter** [comment]
If the fix is signaled (eg: orange text) then yes!

**Comment by XyLoNaMiyX** [comment]
Trying to think how this could fit into the UI. How about the "linter" running right when you try to change to another string, if it finds anything it should warn you about, then it saves your action ("Next", "Previous", whatever), hides the prev/save/next buttons and instead shows "Cancel", "Apply fix" and "Dismiss". Above these new buttons some [warning icon](https://material.io/icons/#ic_warning) and a text hinting what should be done.

**Comment by licaon-kter** [comment]
Not sure what lint checks apply here.

Another issue is letter case, eg: start of text or in the middle after a dot

**Comment by XyLoNaMiyX** [comment]
> Not sure what lint checks apply here.

To be honest I don't think I understand what a linter is. I thought it was something that checked for common issues or mistakes like variables being unused or the access modifier being lower? Anyway, just a fancy word for the "guard" class.

**Comment by XyLoNaMiyX** [comment]
Probably consider [this](https://github.com/LonamiWebs/Stringlate/pull/88#issuecomment-278867264) for French too.

**Comment by mquinson** [comment]
This editor implements a lot of such verification. You could maybe reuse some code here?
https://github.com/vslavik/poedit

**Comment by Lonami** [comment]
I've never worked with poedit before so I don't know how its verification works, and neither where I should look for such checks 😅 But maybe it's something to take into account thanks @mquinson

**Comment by gsantner** [comment]
Seems harder to get into poedit's code rather than writing this things new in java :D

**Comment by mquinson** [comment]
That's very possible, I never looked at the source code of poedit. If so, maybe you could use poedit code as a list of checks to reimplement in stringlate?

**Comment by gsantner** [comment]
As translator always helpful :D. Feel free to start contributing too 😄
